src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+using AsyncCompletionData = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.AsyncCompletion
+{
+    internal sealed class CommitManager : IAsyncCompletionCommitManager
+    {
+        private static readonly IEnumerable<char> s_commitChars = ImmutableArray.Create(
+            ' ', '{', '}', '[', ']', '(', ')', '.', ',', ':',
+            ';', '+', '-', '*', '/', '%', '&', '|', '^', '!',
+            '~', '=', '<', '>', '?', '@', '#', '\'', '\"', '\\');
+
+        private static readonly AsyncCompletionData.CommitResult CommitResultUnhandled =
+            new AsyncCompletionData.CommitResult(isHandled: false, AsyncCompletionData.CommitBehavior.None);
+
+        public IEnumerable<char> PotentialCommitCharacters => s_commitChars;
+
+        /// <summary>
+        /// The method performs a preliminarily filtering of commit availability.
+        /// In case of a doubt, it should respond with true.
+        /// We will be able to cancel later in TryCommit based on VSCompletionItem item, e.g. based on CompletionItemRules.
+        /// </summary>
+        public bool ShouldCommitCompletion(
+            IAsyncCompletionSession session, 
+            SnapshotPoint location, 
+            char typedChar, 
+            CancellationToken cancellationToken)
+            => s_commitChars.Contains(typedChar);
+
+        public AsyncCompletionData.CommitResult TryCommit(
+            IAsyncCompletionSession session, 
+            ITextBuffer subjectBuffer,
+            VSCompletionItem item, 
+            char typeChar, 
+            CancellationToken cancellationToken)
+        {
```

Comment:
does this method have a threading expectation?  if so, we should doc it and put in an appropriate assert.  if not, then doing subjectBuffer.CurrentSnapshot is not safe.

Reply:
Thank you! Per Editor documentation, it is called from the UI thread. Need to find how to check better if we're in the thread. 

---
In reply to: [222519489](https://github.com/dotnet/roslyn/pull/29016#discussion_r222519489) [](ancestors = 222519489)

Reply:
See https://github.com/dotnet/roslyn/pull/29016/files#r227947931 for a proposed approach.